File: docs/Painter/ph2d_wet_paint/js/engine/colorops.js
// Color math (spec section 14): HSV<->RGB for the UI wheel, sRGB<->linear,
// and Kubelka-Munk single-constant subtractive mixing (opt-in).
//
// Engine colors are ALWAYS 0..255 floats and are never quantized to bytes:
// a wet cell can re-mix thousands of times, and a byte round-trip on every
// mix drifts washes toward black.

/** h in [0,360), s,v in [0,1] -> [r,g,b] 0..255 floats. */
export function hsvToRgb(h, s, v) {
  const c = v * s;
  const hp = (((h % 360) + 360) % 360) / 60;
  const x = c * (1 - Math.abs((hp % 2) - 1));
  let r = 0, g = 0, b = 0;
  if (hp < 1) { r = c; g = x; } else if (hp < 2) { r = x; g = c; }
  else if (hp < 3) { g = c; b = x; } else if (hp < 4) { g = x; b = c; }
  else if (hp < 5) { r = x; b = c; } else { r = c; b = x; }
  const m = v - c;
  return [(r + m) * 255, (g + m) * 255, (b + m) * 255];
}

/** r,g,b 0..255 -> [h 0..360, s 0..1, v 0..1]. */
export function rgbToHsv(r, g, b) {
  const rn = r / 255, gn = g / 255, bn = b / 255;
  const max = Math.max(rn, gn, bn), min = Math.min(rn, gn, bn);
  const d = max - min;
  let h = 0;
  if (d > 0) {
    if (max === rn) h = 60 * (((gn - bn) / d) % 6);
    else if (max === gn) h = 60 * ((bn - rn) / d + 2);
    else h = 60 * ((rn - gn) / d + 4);
  }
  if (h < 0) h += 360;
  return [h, max === 0 ? 0 : d / max, max];
}

/** Standard sRGB EOTF, c in [0,1] -> linear [0,1]. */
export function srgbToLinear(c) {
  return c <= 0.04045 ? c / 12.92 : Math.pow((c + 0.055) / 1.055, 2.4);
}

/** Inverse EOTF, linear [0,1] -> sRGB [0,1]. */
export function linearToSrgb(c) {
  return c <= 0.0031308 ? c * 12.92 : 1.055 * Math.pow(c, 1 / 2.4) - 0.055;
}

// ---------------------------------------------------------------------------
// Kubelka-Munk single-constant mixing.
// A reflectance R maps to an absorption/scattering ratio K/S = (1-R)^2 / 2R.
// Pigment mixtures are LINEAR in K/S, so we convert both colors, lerp the
// ratios, and invert: R = 1 + KS - sqrt(KS^2 + 2 KS). Reflectance is floored
// at 1/255 so pure black cannot make KS blow up.
// ---------------------------------------------------------------------------

const R_FLOOR = 1 / 255;

function ksOfReflectance(r) {
  const rr = r < R_FLOOR ? R_FLOOR : r;
  return ((1 - rr) * (1 - rr)) / (2 * rr);
}

function reflectanceOfKs(ks) {
  return 1 + ks - Math.sqrt(ks * ks + 2 * ks);
}

/**
 * Mix one linear-light channel value (0..1) toward another by weight w in K/S
 * space. Both endpoints exact: w=0 -> dst, w=1 -> src.
 */
export function kmMixChannelLinear(dstLin, srcLin, w) {
  const ks = (1 - w) * ksOfReflectance(dstLin) + w * ksOfReflectance(srcLin);
  return reflectanceOfKs(ks);
}

/**
 * Mix two engine colors (0..255 float channels) by weight w using K-M.
 * Full float pipeline: bytes-float -> sRGB -> linear -> K/S lerp -> back.
 * Writes into out[0..2].
 */
export function kmMixColor(dr, dg, db, sr, sg, sb, w, out) {
  out[0] = linearToSrgb(kmMixChannelLinear(srgbToLinear(dr / 255), srgbToLinear(sr / 255), w)) * 255;
  out[1] = linearToSrgb(kmMixChannelLinear(srgbToLinear(dg / 255), srgbToLinear(sg / 255), w)) * 255;
  out[2] = linearToSrgb(kmMixChannelLinear(srgbToLinear(db / 255), srgbToLinear(sb / 255), w)) * 255;
}

/** Plain linear-RGB lerp of two engine colors by weight w, into out[0..2]. */
export function plainMixColor(dr, dg, db, sr, sg, sb, w, out) {
  out[0] = dr + (sr - dr) * w;
  out[1] = dg + (sg - dg) * w;
  out[2] = db + (sb - db) * w;
}

/**
 * The engine routes every pigment-color blend through this indirection.
 * `useKm` is the "pigment mixing (K-M)" experimental checkbox; default OFF
 * keeps the plain composites of spec sections 6/10/11.
 */
export function makeColorMixer(useKm) {
  return useKm ? kmMixColor : plainMixColor;
}

/**
 * K-M weighted mean of up to 4 engine colors (used by the advection's
 * incoming-color mean when pigment mixing is ON): mixtures are linear in
 * K/S, so we average the corners' K/S ratios with the pull weights.
 * colors = flat array [r0,g0,b0, r1,g1,b1, ...] (0..255), weights sum > 0.
 */
export function kmWeightedMeanColor(colors, weights, count, invTotal, out) {
  for (let ch = 0; ch < 3; ch++) {
    let ks = 0;
    for (let k = 0; k < count; k++) {
      ks += weights[k] * ksOfReflectance(srgbToLinear(colors[k * 3 + ch] / 255));
    }
    out[ch] = linearToSrgb(reflectanceOfKs(ks * invTotal)) * 255;
  }
}

/**
 * K-M glaze stack of one pigment layer (linear reflectance rTop, coverage a)
 * over a backdrop (linear reflectance rBot). Energy-bounded: a=1 -> rTop,
 * a=0 -> rBot. Guard the denominator against rTop*rBot ~ 1.
 */
export function kmGlazeChannelLinear(rBot, rTop, a) {
  const rt = a * rTop;
  const tt = 1 - a;
  const den = 1 - rt * rBot;
  const r = rt + (tt * tt * rBot) / (den < 1e-6 ? 1e-6 : den);
  return r < 0 ? 0 : r > 1 ? 1 : r;
}

File: crates/ph2d-wet-paint/src/colorops.rs
//! Color math (port of `colorops.js`, SPEC §14): HSV<->RGB for UI, sRGB<->
//! linear, and Kubelka–Munk single-constant subtractive mixing (opt-in).
//!
//! Engine colors are ALWAYS 0..255 floats and are never quantized to bytes:
//! a wet cell can re-mix thousands of times, and a byte round-trip on every
//! mix drifts washes toward black.
//!
//! The sRGB transfer runs only when the K–M checkbox is ON or the glaze stack
//! renders — never in the default hot path (HR-5) — but when it does run it
//! runs 9-15 times per cell, which made it the whole cost of both
//! EXPERIMENTAL knobs. It lives in [`transfer`], node-tabulated from `libm`
//! and interpolated with IEEE basic operations only: same cross-OS
//! bit-identity, ~11x the speed. Read that module before touching it.

/// h in [0,360), s,v in [0,1] -> (r,g,b) 0..255 floats.
pub fn hsv_to_rgb(h: f64, s: f64, v: f64) -> [f64; 3] {
    let c = v * s;
    let hp = (((h % 360.0) + 360.0) % 360.0) / 60.0;
    let x = c * (1.0 - ((hp % 2.0) - 1.0).abs());
    let (mut r, mut g, mut b) = (0.0, 0.0, 0.0);
    if hp < 1.0 {
        r = c;
        g = x;
    } else if hp < 2.0 {
        r = x;
        g = c;
    } else if hp < 3.0 {
        g = c;
        b = x;
    } else if hp < 4.0 {
        g = x;
        b = c;
    } else if hp < 5.0 {
        r = x;
        b = c;
    } else {
        r = c;
        b = x;
    }
    let m = v - c;
    [(r + m) * 255.0, (g + m) * 255.0, (b + m) * 255.0]
}

/// r,g,b 0..255 -> (h 0..360, s 0..1, v 0..1).
pub fn rgb_to_hsv(r: f64, g: f64, b: f64) -> [f64; 3] {
    let rn = r / 255.0;
    let gn = g / 255.0;
    let bn = b / 255.0;
    let max = rn.max(gn).max(bn);
    let min = rn.min(gn).min(bn);
    let d = max - min;
    let mut h = 0.0;
    if d > 0.0 {
        if max == rn {
            h = 60.0 * (((gn - bn) / d) % 6.0);
        } else if max == gn {
            h = 60.0 * ((bn - rn) / d + 2.0);
        } else {
            h = 60.0 * ((rn - gn) / d + 4.0);
        }
    }
    if h < 0.0 {
        h += 360.0;
    }
    [h, if max == 0.0 { 0.0 } else { d / max }, max]
}

pub use ph2d_pigment::transfer;

// The standard sRGB EOTF and its inverse, plus the 0..255-domain doors the
// K–M and glaze sites use. ONE door each: every experimental site goes through
// these, and the tables live behind them so no call site can pick a different
// transfer than its neighbour.
pub use ph2d_pigment::{ks_of_srgb255, linear_to_srgb, srgb_to_linear, srgb255_of_linear};

// ── Kubelka–Munk: a lei MUDOU-SE para a folha [`ph2d_pigment`] (ordem do dono 2026-09-20, «os três
// meios passam a misturar igual»). Ela não mudou uma linha de matemática — mudou de ENDEREÇO, para
// que o Digital e a Aquarela leiam esta e não uma cópia. Este motor delega, logo continua
// byte-idêntico; ⛔ nada de reescrever `ColorMix` aqui, que é a segunda cópia que diverge. ──
pub use ph2d_pigment::{ColorMix, reflectance_of_ks};

/// K–M weighted mean of up to 4 engine colors (the advection's incoming-color
/// mean when pigment mixing is ON): mixtures are linear in K/S, so average
/// the corners' K/S ratios with the pull weights.
/// `colors` = flat [r0,g0,b0, r1,g1,b1, ...] (0..255); `inv_total` = 1/Σw.
pub fn km_weighted_mean_color(
    colors: &[f64; 12],
    weights: &[f64; 4],
    count: usize,
    inv_total: f64,
    out: &mut [f64; 3],
) {
    for ch in 0..3 {
        let mut ks = 0.0;
        for k in 0..count {
            // One door per corner per channel: the `/255` rescale rides inside
            // the table's index instead of costing a division each.
            ks += weights[k] * ks_of_srgb255(colors[k * 3 + ch]);
        }
        out[ch] = srgb255_of_linear(reflectance_of_ks(ks * inv_total));
    }
}

/// K–M glaze stack of one pigment layer (linear reflectance r_top, coverage a)
/// over a backdrop (linear reflectance r_bot). Energy-bounded: a=1 -> r_top,
/// a=0 -> r_bot. Guard the denominator against r_top * r_bot ~ 1.
#[inline]
pub fn km_glaze_channel_linear(r_bot: f64, r_top: f64, a: f64) -> f64 {
    let rt = a * r_top;
    let tt = 1.0 - a;
    let den = 1.0 - rt * r_bot;
    let r = rt + (tt * tt * r_bot) / if den < 1e-6 { 1e-6 } else { den };
    r.clamp(0.0, 1.0)
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn hsv_round_trips() {
        let [r, g, b] = hsv_to_rgb(210.0, 0.76, 0.82);
        let [h, s, v] = rgb_to_hsv(r, g, b);
        assert!((h - 210.0).abs() < 0.01 && (s - 0.76).abs() < 0.01 && (v - 0.82).abs() < 0.01);
    }

    #[test]
    fn km_mix_endpoints_are_bit_exact_at_any_colour() {
        // A weight of 0 returns dst and a weight of 1 returns src, to the
        // BIT, for every colour — including BELOW the 1/255 reflectance floor,
        // where the K/S round trip provably cannot recover the input (every
        // colour there shares one K/S). The endpoints do not go through K/S at
        // all, so the floor is irrelevant to them; asserting equality rather
        // than a tolerance is what pins that, and it is what keeps a pass that
        // deposits nothing from tinting paint it did not touch.
        let mut out = [0.0; 3];
        for (d, s) in [
            ([200.0, 90.0, 60.0], [40.0, 120.0, 180.0]),
            ([3.0, 1.0, 0.0], [255.0, 254.0, 250.0]), // both ends, floor included
        ] {
            ColorMix::Km.mix(d[0], d[1], d[2], s[0], s[1], s[2], 0.0, &mut out);
            assert_eq!(out, d, "w=0 must return dst unchanged");
            ColorMix::Km.mix(d[0], d[1], d[2], s[0], s[1], s[2], 1.0, &mut out);
            assert_eq!(out, s, "w=1 must return src exactly");
        }
    }

    #[test]
    fn km_yellow_plus_blue_makes_green() {
        // The signature of subtractive mixing: yellow + blue reads GREEN —
        // the green channel dominates harder than in an additive lerp (the
        // absolute values are darker; subtractive mixing absorbs).
        let mut km = [0.0; 3];
        ColorMix::Km.mix(250.0, 205.0, 40.0, 10.0, 70.0, 150.0, 0.5, &mut km);
        let mut plain = [0.0; 3];
        ColorMix::Plain.mix(250.0, 205.0, 40.0, 10.0, 70.0, 150.0, 0.5, &mut plain);
        assert!(km[1] > km[0] && km[1] > km[2], "K-M mid not green: {km:?}");
        let km_dominance = km[1] / km[0].max(km[2]);
        let plain_dominance = plain[1] / plain[0].max(plain[2]);
        assert!(
            km_dominance > plain_dominance,
            "K-M green dominance {km_dominance:.3} !> plain {plain_dominance:.3}"
        );
    }

    #[test]
    fn glaze_is_energy_bounded() {
        assert!((km_glaze_channel_linear(0.3, 0.8, 1.0) - 0.8).abs() < 1e-12);
        assert!((km_glaze_channel_linear(0.3, 0.8, 0.0) - 0.3).abs() < 1e-12);
    }
}
